### ISSUE TITLE
Distributed training with keras does not seem to synchronize models of each node.

### ISSUE THREAD

**Issue by y-rok** [horovod/horovod#1512]
**Environment:**
1. Framework: Keras
2. Framework version:
3. Horovod version:0.18.2
4. MPI version:
5. CUDA version:10.0
6. NCCL version:
7. Python version:3.7.5
8. OS and version:linux ubuntu 18.04
9. GCC version:4.8.5

Hello.

I've written codes for resnet distributed training with keras.

When distributed training, the gradients of all nodes are averaged at each step before updating model.
Therefore, test loss and accuracy of all models must be same after training.

However, when i run below codes with 2 gpu (2 processes), test loss and accuracy of two nodes are different.

Please let me know why this problem happens.

`

    # Horovod: initialize Horovod.
    hvd.init()

    # Horovod: pin GPU to be used to process local rank (one GPU per process)
    config = tf.ConfigProto()
    config.gpu_options.allow_growth = True
    config.gpu_options.visible_device_list = str(hvd.local_rank())
    K.set_session(tf.Session(config=config))

    # batch_size_per_node = int(BATCH_SIZE/hvd.size())

    # Load the CIFAR10 data.
    (x_train, y_train), (x_test, y_test) = cifar10.load_data()

    # Input image dimensions.
    input_shape = x_train.shape[1:]

    # Normalize data.
    x_train = x_train.astype('float32') / 255
    x_test = x_test.astype('float32') / 255

    # Subtracting pixel mean improves accuracy
    x_train_mean = np.mean(x_train, axis=0)
    x_test_mean = np.mean(x_test, axis=0)
    x_train -= x_train_mean
    x_test -= x_test_mean

    # Convert class vectors to binary class matrices.
    y_train = keras.utils.to_categorical(y_train, NUM_CLASSES)
    y_test = keras.utils.to_categorical(y_test, NUM_CLASSES)

    model = resnet_v1(input_shape=input_shape, depth=DEPTH)

    opt = keras.optimizers.Adam(0.01*hvd.size())
    # Add Horovod Distributed Optimizer
    opt = hvd.DistributedOptimizer(opt)

    callbacks = [
        # Horovod: broadcast initial variable states from rank 0 to all other processes.
        # This is necessary to ensure consistent initialization of all workers when
        # training is started with random weights or restored from a checkpoint.
        hvd.callbacks.BroadcastGlobalVariablesCallback(0),
    ]

    model.compile(loss=keras.losses.categorical_crossentropy,
                  optimizer=opt,
                  metrics=['accuracy'])
    model.summary()

    # Horovod: save checkpoints only on worker 0 to prevent other workers from corrupting them.
    # if hvd.rank() == 0:
    #     callbacks.append(keras.callbacks.ModelCheckpoint('./checkpoint-{epoch}.h5'))

    current_time = time.time()

    model.fit(x_train, y_train,
              batch_size=BATCH_SIZE,
              epochs=int(EPOCHS/hvd.size()),
              validation_data=(x_test, y_test),
              callbacks=callbacks)

    print("Elapsed time = ",time.time()-current_time)

    # Score trained model.
    scores = model.evaluate(x_test, y_test, verbose=0)
    print('Test loss:', scores[0])
    print('Test accuracy:', scores[1])
`

**Comment by tgaddair** [comment]
Hey @y-rok, how different are the models?  In some cases, small divergences can occur due to numerical non-determinism in the training process.  If this is causing an issue with convergence, you can modify the `hvd.callbacks.BroadcastGlobalVariablesCallback(0)` callback to run at the end of each epoch.  That will force all model replicas across all workers to share the same state.

**Comment by y-rok** [comment]
@tgaddair  Thanks for your reply :)
Could you explain me the meaning of "numerical non-determinism"?

The test losses of nodes are slightly different.
As i know, the gradients are averaged across nodes before updating model at each step.
So I think the models of nodes should be same.

**Comment by tgaddair** [comment]
Hey @y-rok, to clarify, there are some operations (for example, floating-point summation) which occur in different orders on different devices.  Because floating-point summation is not exactly precise, the order of summation can slightly change the final results, leading to different gradients on different workers, in some cases.  I would suggest trying to synchronize model state more frequently and see if that solves the issue.

**Comment by y-rok** [comment]
@tgaddair Is there any way to check if gradients are averaged before updating model? I am curious if there is log i can check.

**Comment by tgaddair** [comment]
Hey @y-rok, in TensorFlow you can look at the graph.  There should be a Horovod Allreduce op in the graph that occurs before gradients are applied to update the model.  You can also check that allreduce is occurring on all the gradients by creating a [timeline](https://github.com/horovod/horovod/blob/master/docs/timeline.rst).

**Comment by y-rok** [comment]
@tgaddair Thank you! i will try it
